src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs

```diff
@@ -0,0 +1,139 @@
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = CSharpSyntaxContext.CreateContext(workspace, semanticModel, position, cancellationToken);
+
+            if (!syntaxContext.IsAwaitKeywordContext(position))
+            {
+                return;
+            }
+
+            var method = syntaxContext.TargetToken.GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            var shouldMakeContainerAsync = method is not null && !method.GetModifiers().Any(SyntaxKind.AsyncKeyword);
+            context.AddItem(GetCompletionItem(shouldMakeContainerAsync));
+
+            return;
+
+            static CompletionItem GetCompletionItem(bool shouldMakeContainerAsync)
+                => CommonCompletionItem.Create(
+                       displayText: SyntaxFacts.GetText(SyntaxKind.AwaitKeyword),
+                       displayTextSuffix: "",
+                       rules: CompletionItemRules.Default,
+                       Glyph.Keyword,
+                       description: RecommendedKeyword.CreateDisplayParts(SyntaxFacts.GetText(SyntaxKind.AwaitKeyword), string.Empty),
+                       inlineDescription: shouldMakeContainerAsync ? CSharpFeaturesResources.Make_container_async : null,
+                       isComplexTextEdit: shouldMakeContainerAsync);
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            // IsComplexTextEdit is true when we want to add async to the container.
+            if (!item.IsComplexTextEdit)
+            {
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var declaration = root.FindToken(item.Span.Start).GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            if (declaration is null)
+            {
+                // We already check that in ProvideCompletionsAsync above.
+                Debug.Assert(false, "Expected non-null value for declaration.");
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            var documentWithAsyncModifier = document.WithSyntaxRoot(root.ReplaceNode(declaration, AddAsyncModifier(declaration)));
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var builder);
+
+            builder.AddRange(await documentWithAsyncModifier.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false));
```

Comment:
this part is actually a bit terrifying.  tehre is no invariant that GetTextChanges returns good edits.  Instead of going this route, i would instead either:

1. just literally hardcode in the logic for determining the span where `async ` should go.
2. use AddAsyncModifier, then directly check the new declaration for the `async` token and figure out the span from taht

Either of those will give a nice narrow text edit that we can depend on.

This is a blocker for this PR currently.

Reply:
Note: This was following what's currently done in another completion provider:

https://github.com/dotnet/roslyn/blob/5624337871b6dd8a2cab70a190878b6a30b2c603/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs#L156

I'll try with either of the approaches you mentioned.

> tehre is no invariant that GetTextChanges returns good edits

I'm curious what "good edits" mean?